### PR TITLE
chore: Task Workers Revamp

### PR DESCRIPTION
**Pull request by jsantell**
### Decoupled Worker Tasks

Worker tasks are now abstracted such that routes now send jobs via a `JobClient` decoupled from where the work is performed. A standalone `noosphere_gateway::jobs::worker_queue` module provides the work distribution logic, with each processor receiving a `GatewayJobContext` that provides access to sphere contexts, IPFS, and name resolvers when processing a job.

* Worker jobs can now queue up additional tasks on success (IPFS Syndication -> Name Record Publish) (https://github.com/subconsciousnetwork/noosphere/issues/606).
* Jobs can be retried upon failure. 
* Some timeout logic has been implemented, but not yet ready.
* Republishing name records is now its own job type. Eliminates the need for the scheduler to resolve the latest name record.

### Minimize external queries

`GatewayManager` and routes were updated to minimize GM queries (https://github.com/subconsciousnetwork/noosphere/issues/753).

In addition to a single, cacheable baseline query for scope, unauthenticated routes now has one less external query, and authenticated routes also has one less external query, with another query moved behind authorization.

Ultimately handler logic *should not* be changed, other than moving Push handler's queueing of name record publishing behind IPFS syndication.

## Caveats/Concerns/Enhancements

* We should do some profiling to measure lock contention across workers where appropriate amongst shared resources like `KuboClient` and the name resolver.
* Job processors may need to be rethought in order to better take advantage of retries e.g. classifying a failure between retriables, invalid state/unretriable, and failed-but-ok-to-lazily-try-again-later.
* ~~Implement timeouts in worker queue.~~ Timeouts now implemented
* Ipfs syndication revision parameter now optional (unused? to confirm)
* Ensure name record provided to PUSH is still valid after Ipfs syndication completes due to record expiry

### CODE REVIEW

**Comment by jsantell** [comment]
Added explicit checks on existence of name records after syncing changes in tests, possibly clearing up a handful of intermittent failures.

Windows builds are having an unrelated issue: https://github.com/subconsciousnetwork/noosphere/pull/805

File: rust/noosphere-gateway/src/single_tenant/job_client.rs [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+/// The number of worker threads to process jobs.
+const WORKER_COUNT: usize = 5;
+/// How many times by default jobs should retry upon failure.
+/// @TODO Consider allowing individual jobs to hint if they
+/// should be retried, versus failures that can be attempted
+/// on next periodic cycle.
+const JOB_RETRIES: usize = 1;
+/// How many seconds before a job is considered broken,
+/// and potentially restarted.
+const TIMEOUT_SECONDS: u64 = 180;
```
Comment (jsantell) [inline rust/noosphere-gateway/src/single_tenant/job_client.rs:30]: Syndicating large histories could be longer than this timeout, and notably, name records' publishing timeframe is currently 2 minutes.

File: rust/noosphere-gateway/src/jobs/worker_queue/queue.rs [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    /// By default, `retries` is set to 1 and `timeout` is 3 minutes.
+    pub fn spawn(
+        worker_count: usize,
+        worker_context: P::Context,
+        retries: Option<usize>,
+        timeout: Option<Duration>,
+    ) -> Self {
+        let (request_tx, request_rx) = unbounded_channel();
+        let handle = Some(Arc::new(tokio::spawn(async move {
+            let workers = WorkerQueueThread::<P>::new(
```
Comment (cdata) [inline rust/noosphere-gateway/src/jobs/worker_queue/queue.rs:35]: The naming and documentation were confusing to read the first time around. `WorkerQueue` sounds like "the thing that has a pool of workers." In fact, the documentation currently reads:

```rs
/// An abstraction around managing several worker threads, and distributing
/// work amongst them.
```

And, there is `WorkerQueueThread`, which sounds like the thing that does work on a thread. But, most of the meaningful management of multiple threads is performed by `WorkerQueueThread`.

Why are these separate constructs? It kind of looks like all that `WorkerQueue` does is thinly wrap `WorkerQueueThread`. Also: can we call `WorkerQueueThread` something closer to what it is? Maybe it's just `WorkerQueue`, maybe it's a `WorkerQueuePool`. It's certainly not a single thread so long as it has a `Vec<Workers<P>>`.

File: rust/noosphere-gateway/src/jobs/worker_queue/worker.rs [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("JobRequest")
+            .field("job", &self.job)
+            .field("attempts", &self.attempt)
+            .finish()
+    }
+}
+
+/// Provides an interface to perform a job via [Processor]
+/// on another thread.
```
Comment (cdata) [inline rust/noosphere-gateway/src/jobs/worker_queue/worker.rs:54]: Can you comment about the usage and lifecycle of a `Worker` in the documentation?

File: rust/noosphere/tests/distributed_basic.rs [hunk 4]
```diff
@@ -292,35 +295,45 @@ async fn traverse_spheres_and_get_incremental_updates_via_noosphere_gateway_via_
...
     for pair in [&pair_1, &pair_2] {
         let name = pair.name.clone();
         let mut ctx = pair.sphere_context().await?;
         ctx.write("my-name", "text/plain", name.as_ref(), None)
             .await?;
         ctx.save(None).await?;
-        ctx.sync().await?;
+        initial_versions.push(ctx.sync().await?);
+    }
+    {
```
Comment (cdata) [inline rust/noosphere/tests/distributed_basic.rs:319]: Why is the extra block needed?

**Review by cdata — CHANGES_REQUESTED** [review]
The bones look excellent, but I have some light opinions about some of the flesh :)

Great work so far!

File: rust/noosphere/tests/distributed_basic.rs (hunk 4 above)
Comment (jsantell) [reply rust/noosphere/tests/distributed_basic.rs:319]: Looks unnecessary, will remove

**Comment by jsantell** [comment]
Updates ([commit](https://github.com/subconsciousnetwork/noosphere/pull/796/commits/8cadcfc18fb9e514ebf7d4426bd97bda06476e11)):
* Changes/cleanups from feedback
* Remove vestigial `revision` in IpfsSyndication job
* Reduce indirection from cloning (e.g. use outer Arcs, not inner)
* s/ContextResolver/SphereContextResolver
* s/NameResolverExt/NameResolverPoller
* s/WorkerQueueThread/WorkerQueueOrchestrator
* More documentation for `worker_queue`

**Review by cdata — APPROVED** [review]
LGTM, thanks for the hard work on this!

**Pull request merged**